### PR TITLE
Remove LCOW (step 5): volumes/mounts: remove LCOW code (alternative)

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to (part of) https://github.com/moby/moby/pull/42170
alternative for https://github.com/moby/moby/pull/42517
closes https://github.com/moby/moby/pull/42517

This alternative to https://github.com/moby/moby/pull/42517 preserves the LCOW parsing code for mounts (in case we will be using LCOW using containerd's LCOW implementation). By default, a native parser is now used (matching the host (and thus container's) platform.

After taking care of the above, tests have also be split to be per-driver, and some refactoring to improve the tests (being able to run them "parallel").

details below:

### volume/mounts: move some code to correct location, and minor linting/formatting

- Remove the windowsparser.HasResource() override, as it was the same on both
  Windows and Linux
- Move the rxLCOWDestination to the lcowParser code
- Move the rwModes variable to a generic (non-platform-specific) file, as it's
  used both for the windowsParser and the linuxParser
- Some minor formatting and linting changes

### volume/mounts: remove "containerOS" argument from NewParser (LCOW code)

This changes mounts.NewParser() to create a parser for the current operatingsystem,
instead of one specific to a (possibly non-matching, in case of LCOW) OS.

With the OS-specific handling being removed, the "OS" parameter is also removed
from `daemon.verifyContainerSettings()`, and various other container-related
functions.

### Split tests per parser

Have tests per-parser, instead of grouping all tests in a single test (with a
test-set for each parser inside that test). All tests have been kept, at the
cost of some duplication in the test-cases that overlapped / were the same
for multiple parsers.

### Add constructors for each parser

This adds constructors for the Linux, Windows, and LCOW, to allow using these
parsers externally, instead of only through `NewParser()`.

### Don't use global variable for fileinfoprovider

The code used a global / package-level `currentFileInfoProvider` that was used
to stub the actual fileInfoProvider during tests. Using a package-level variable
for this could cause issues when trying to run tests in parallel, as well as
making the tests more complicated (tests had to reset the provider in a defer).

Now that we have a constructor for each driver, we can set up the fileinfoprovider
in those constructors, and stub the implementation for specific instances, instead
of globally.

### Use sub-tests, and use gotest.tools

These tests were running many test-cases, which required a lot of glue code to
construct useful errors if "one of those" failed.

Using regular sub-tests instead to make debugging easier, and use gotest.tools
to take care of useful errors.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@tianon ptal if you like this approach better than the other one (i.e., keeping the functionality)

**Comment by thaJeztah** [comment]
😬 guess I need to update the commit message for the first commit (I was happy that it was green ✅ ); can do so if we decide to go the route in this PR.

**Comment by tianon** [comment]
Yeah, I think this makes more sense if we plan to keep LCOW in a different form -- there was some debate on today's maintainers call as to whether that's even something we want, so I guess we need to decide on that before we can really say for sure whether we can just bring down the axe or whether we need to keep all three parsers. :smile:

**Comment by thaJeztah** [comment]
@tianon @AkihiroSuda @cpuguy83 I re-ordered some of the commits, and updated the PR description to explain what this PR does, and how it differs from the other PR.

**Review by cpuguy83 — COMMENTED** [review]
I think it'll be easier to keep the the argument to `NewParser`... at least since we aren't removing this parsing I don't see the benefit in ripping out the parts that enable switching.

**Comment by thaJeztah** [comment]
> I think it'll be easier to keep the the argument to `NewParser`... at least since we aren't removing this parsing I don't see the benefit in ripping out the parts that enable switching.

Do you want to keep all the code that passes it, or just that argument (and by default call it with `runtime.GOOS`)?

Mostly I tried to get rid of all the `if foo == "windows"` and `if container.OS != runtime.GOOS` code that touches every bit of the code.

**Comment by thaJeztah** [comment]
ping @cpuguy83 ^^ could you explain what you wanted to see here?

**Comment by thaJeztah** [comment]
@tianon @cpuguy83 - per our discussion in the maintainers meeting; this LGTY?

File: volume/mounts/lcow_parser.go [hunk 1]
```diff
@@ -7,6 +7,23 @@ import (
...
+	}
+}
+
+// rxLCOWDestination is the regex expression for the mount destination for LCOW
+//
+// Destination (aka container path):
+//    -  Variation on hostdir but can be a drive followed by colon as well
+//    -  If a path, must be absolute. Can include spaces
+//    -  Drive cannot be c: (explicitly checked in code, not RegEx)
+const rxLCOWDestination = `(?P<destination>/(?:[^\\/:*?"<>\r\n]+[/]?)*)`
```
Comment (tianon) [inline volume/mounts/lcow_parser.go:25]: This isn't new, so I'm not suggesting we change it now, but shouldn't this (and the friends it moved away from) be `regexp.MustCompile`'d?

**Review by tianon — APPROVED** [review]
LGTM

File: volume/mounts/lcow_parser.go (hunk 1 above)
Comment (thaJeztah) [reply volume/mounts/lcow_parser.go:25]: Yes, probably makes sense to do so (help me remind in case I forget)

Comment (tianon) [reply volume/mounts/lcow_parser.go:25]: https://github.com/moby/moby/issues/42667 :smirk: :innocent:

**Comment by justincormack** [comment]
I don't see any evidence that there will be a containerd LCOW implementation in the foreseeable future, so why not remove the code for now?

**Comment by thaJeztah** [comment]
@justincormack there was some back-and-forth on that. That said, the remaining code in this package is fairly limited, and is just what's in https://github.com/thaJeztah/docker/blob/remove_lcow_step5_alternative/volume/mounts/lcow_parser.go (and some tests in the `lcow_parser_test.go`)

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Pull request merged**